### PR TITLE
Revised nullable (value) types articles

### LINKED ISSUE #10141: The article needs some modification about nullable reference types
The article said that:

> Nullable types represent value-type variables that can be assigned the null value. You cannot create a nullable type based on a reference type. (Reference types already support the null value.)

Now in the incoming C# 8.0, a reference type can be a nullable type.
The article above needs some modification, right?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ab99a9ea-4645-92cd-723a-4a38fbee371e
* Version Independent ID: 652ea928-6ded-800c-6bed-5b1f3f67fe3f
* Content: [Nullable types - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/)
* Content Source: [docs/csharp/programming-guide/nullable-types/index.md](https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/nullable-types/index.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by pkulikov**
Fixes #10141
Contributes to #13421 

Hiding whitespace changes may simplify the review.

@BillWagner while working on that, I've noticed that the [overview](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/) and [Using nullable value types](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/using-nullable-types) share a lot. I suggest to combine all **three** articles about the nullable value types and move them from the programming guide to the **one** article in the language reference (*builtin-types* folder). What do you think?

### CODE REVIEW

**Comment by BillWagner** [comment]
@pkulikov 

> I've noticed that the overview and Using nullable value types share a lot. I suggest to combine all three articles about the nullable value types and move them from the programming guide to the one article in the language reference (builtin-types folder). What do you think?

I do like that plan.  Do you want to create an issue for that work?

File: docs/csharp/programming-guide/nullable-types/using-nullable-types.md [hunk 1]
```diff
@@ -1,85 +1,88 @@
...
-# Using nullable types (C# Programming Guide)
+# Using nullable value types (C# Programming Guide)
 
-Nullable types are types that represent all the values of an underlying value type `T`, and an additional [null](../../language-reference/keywords/null.md) value. For more information, see the [Nullable types](index.md) topic.
+Nullable value types are types that represent all the values of an underlying value type `T`, and an additional [null](../../language-reference/keywords/null.md) value. For more information, see the [Nullable value types](index.md) topic.
+
+> [!NOTE]
+> C# 8.0 introduces the nullable reference types feature. For more information, see [Nullable reference types](../../nullable-references.md). The nullable value types are available starting with C# 2.
+
+You can refer to a nullable type in any of the following forms: `Nullable<T>` or `T?`. These two forms are interchangeable.
```
Comment (BillWagner) [inline docs/csharp/programming-guide/nullable-types/using-nullable-types.md]: The form `Nullable<T>` is only valid for value types, not reference types. We should reword this sentence.

File: docs/csharp/programming-guide/nullable-types/using-nullable-types.md [hunk 2]
```diff
@@ -1,85 +1,88 @@
...
 
-You can refer to a nullable type in any of the following forms: `Nullable<T>` or `T?`. These two forms are interchangeable.  
-  
 ## Declaration and assignment
 
-As a value type can be implicitly converted to the corresponding nullable type, you assign a value to a nullable type as you would for its underlying value type. You also can assign the `null` value.  For example:
-  
+As a value type can be implicitly converted to the corresponding nullable type, you assign a value to a nullable type as you would for its underlying value type. You also can assign the `null` value. For example:
```
Comment (BillWagner) [inline docs/csharp/programming-guide/nullable-types/using-nullable-types.md]:
```suggestion
As a value type can be implicitly converted to the corresponding nullable value type, you assign a value to a nullable type as you would for its underlying value type. You also can assign the `null` value. For example:
```

**Review by BillWagner — CHANGES_REQUESTED** [review]
Overall, this is great work @pkulikov 

I had a few pieces of feedback, and one discussion item to resolve. 

Thanks again for all your work on our docs!

**Comment by pkulikov** [comment]
> I do like that plan. Do you want to create an issue for that work?

@BillWagner created: #14726

File: docs/csharp/how-to/safely-cast-using-pattern-matching-is-and-as-operators.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ The following code demonstrates the pattern matching `is` statement. It contains
 
 The preceding sample demonstrates a few features of pattern matching syntax. The `if (a is Mammal m)` and `if (o is Mammal m)` statements combine the test with an initialization assignment. The assignment occurs only when the test succeeds. The variable `m` is only in scope in the embedded `if` statement where it has been assigned. You cannot access `m` later in the same method. Try it in the interactive window.
 
-You can also use the same syntax for testing if a [nullable type](../programming-guide/nullable-types/index.md) has a value, as shown in the following sample code:
+You can also use the same syntax for testing if a [nullable value type](../programming-guide/nullable-types/index.md) has a value, as shown in the following sample code:
```
[the remark this answers is not in the document]
Comment (pkulikov) [reply docs/csharp/how-to/safely-cast-using-pattern-matching-is-and-as-operators.md]: > "value" shouldn't be added in "nullable type" above.

Then link should be either removed or updated. However, I would keep the "value" word in, as the start of the sentence ("You can also use...") implies that this syntax is not exclusive to nullable value types. All in all, the following code

```csharp
string foo = null;
if (foo is string bar)
    Console.WriteLine(bar);
else
    Console.WriteLine("missing value");
```

can be used in C# 8.0 and earlier versions as well. It looks to be fine with both nullable and non-nullable reference types. That said, I think the paragraph *after* the example can be made more generic and explicitly mention reference types.

[the remark this answers is not in the document]
Comment (pkulikov) [reply docs/csharp/how-to/safely-cast-using-pattern-matching-is-and-as-operators.md]: @BillWagner I've updated the paragraph after the example. What do you think about the latest commit?

[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/how-to/safely-cast-using-pattern-matching-is-and-as-operators.md]: This looks great! I'll approve this and :shipit: now.

**Review by BillWagner — APPROVED** [review]

**Pull request merged**